intellij-plugin/src/main/kotlin/com/apollographql/ijplugin/inspection/ApolloMissingGraphQLDefinitionImportInspection.kt

```diff
@@ -0,0 +1,208 @@
+    private val definitions: List<GQLDefinition>,
+    private val definitionsUrl: String,
+) : LocalQuickFix {
+  override fun getName() = ApolloBundle.message("inspection.missingGraphQLDefinitionImport.quickFix", typeKind, "'$elementName'")
+  override fun getFamilyName() = name
+
+  override fun availableInBatchMode() = false
+  override fun generatePreview(project: Project, previewDescriptor: ProblemDescriptor): IntentionPreviewInfo = IntentionPreviewInfo.EMPTY
+
+  override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+    if (!IntentionPreviewUtils.isIntentionPreviewActive()) project.telemetryService.logEvent(TelemetryEvent.ApolloIjMissingGraphQLDefinitionImportQuickFix())
+
+    val element = descriptor.psiElement.parentOfType<GraphQLDirective>(withSelf = true)!!
+    val schemaFiles = element.schemaFiles()
+    val linkDirective = schemaFiles.flatMap { it.linkDirectives(definitionsUrl) }.firstOrNull()
+
+    // Special case: also add a @catch directive to the schema if we're importing @catch
+    val catchDirectiveSchemaExtension = if (element.name == CATCH && schemaFiles.flatMap { it.schemaCatchDirectives() }.isEmpty()) {
+      createCatchDirectiveSchemaExtension(project)
+    } else {
+      null
+    }
+
+    if (linkDirective == null) {
+      val linkDirectiveSchemaExtension = createLinkDirectiveSchemaExtension(project, setOf(element.nameForImport), definitions, definitionsUrl)
+      val extraSchemaFile = schemaFiles.firstOrNull { it.name == "extra.graphqls" }
+      if (extraSchemaFile == null) {
+        val fileText = linkDirectiveSchemaExtension.text + catchDirectiveSchemaExtension?.let { "\n\n" + it.text }.orEmpty()
+        GraphQLElementFactory.createFile(project, fileText).also {
+          // Save the file to the project
+          it.name = "extra.graphqls"
+          schemaFiles.first().containingDirectory!!.add(it)
+
+          // There's a new schema file, reload the configuration
+          project.gradleToolingModelService.triggerFetchToolingModels()
+        }
+      } else {
+        extraSchemaFile.add(GraphQLElementFactory.createWhiteSpace(project, "\n\n"))
+        extraSchemaFile.add(linkDirectiveSchemaExtension)
+        catchDirectiveSchemaExtension?.let {
+          extraSchemaFile.add(GraphQLElementFactory.createWhiteSpace(project, "\n\n"))
+          extraSchemaFile.add(it)
+        }
+      }
+    } else {
+      val extraSchemaFile = linkDirective.containingFile
+      val importedNames = buildSet {
+        addAll(linkDirective.arguments!!.argumentList.firstOrNull { it.name == "import" }?.value?.cast<GraphQLArrayValue>()?.valueList.orEmpty().map { it.text.unquoted() })
+        add(element.nameForImport)
+      }
+      linkDirective.replace(createLinkDirective(project, importedNames, definitions, definitionsUrl))
+      catchDirectiveSchemaExtension?.let {
+        extraSchemaFile.add(GraphQLElementFactory.createWhiteSpace(project, "\n\n"))
+        extraSchemaFile.add(it)
```

Comment:
Can this be added at the top of the file instead, like other imports

Reply:
Good point 👍

with df43394c535fc36c5d4609126bcb78069e5485da:
https://github.com/apollographql/apollo-kotlin/assets/372852/5854bd88-5152-49ba-86c4-728a88d4ad52